### ISSUE TITLE
Error: Only one input size may be -1, not both 0 and 1

### ISSUE THREAD

**Issue by Mochengvia** [tensorflow/tensorflow#76031]
### Issue type

Support

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.17.0

### Custom code

Yes

### OS platform and distribution

Windows 10 LTSC

### Python version

3.12.6

### Current behavior?

When I used tfrecord file to generate the dataset and fit the model, I encountered the error 'Only one input size may be -1, not both 0 and 1'.
 I'm not sure if this issue is caused by a TensorFlow bug, as the main code works fine when I use local image and JSON label files for training.
If there is any way to resolve this issue, please let me know.

### Standalone code to reproduce the issue

```shell
https://colab.research.google.com/drive/1ntVuMurkZ4nmg-owgzYOd9hLA596ZwyE?usp=sharing
```

### Relevant log output

```shell
Total params: 35,040,706 (133.67 MB)
Trainable params: 35,040,706 (133.67 MB)
Non-trainable params: 0 (0.00 B)
Epoch 1/10
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1726724161.327751   51040 service.cc:146] XLA service 0x26fbeda47c0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1726724161.328263   51040 service.cc:154]   StreamExecutor device (0): Host, Default Version
I0000 00:00:1726724161.340495   51040 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.
2024-09-19 13:36:01.640915: W tensorflow/core/framework/op_kernel.cc:1840] OP_REQUIRES failed at reshape_op.h:65 : INVALID_ARGUMENT: 
Only one input size may be -1, not both 0 and 1
2024-09-19 13:36:01.643767: I tensorflow/core/framework/local_rendezvous.cc:404] Local rendezvous is aborting with status: INVALID_ARGUMENT: Only one input size may be -1, not both 0 and 1
         [[{{function_node __inference_one_step_on_data_2728}}{{node functional_1/flatten_1/Reshape}}]]
```

**Comment by Mochengvia** [comment]
After repeated attempts, I found that replacing `tf.keras.layers.Flatten` with `tf.keras.layers.GlobalAveragePooling2D` can solve the problem, although I am not sure what happened. 
The dimensions of the shape of the input are fixed.

**Comment by tilakrayal** [comment]
@Mochengvia,
I was facing a different issue/error while executing the mentioned code. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/03a0bf18a168a244fb522a96fafa9421/untitled2115.ipynb) and provide the complete dependencies which helps to debug the issue. Thank you!

**Comment by Mochengvia** [comment]
@tilakrayal 
Sorry, I forgot to upload my TFRecord file. I’ve updated the code, so you can now run it directly and see the results.
https://colab.research.google.com/drive/1ntVuMurkZ4nmg-owgzYOd9hLA596ZwyE?usp=sharing

**Comment by tilakrayal** [comment]
@Mochengvia,
Tensorflow 2.17 contains Keras3.0 which might be the reason for the error. Could you please try to install **tf-keras** i.e., keras2.0 which the code was executed without any issues/error. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/a8c84c6b9375db93cbb7021679083c95/untitled2142.ipynb).

```python
!pip install tf-keras
import tf_keras as keras
```

Thank you!

**Comment by Mochengvia** [comment]
it works well, thanks.
